src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -878,7 +878,30 @@ private static bool IsGlobalAttributeTarget(SyntaxToken token)
 
         private bool IsPossibleAttributeDeclaration()
         {
-            return this.CurrentToken.Kind == SyntaxKind.OpenBracketToken;
+            // Have to at least start with `[` to be an attribute
+            if (this.CurrentToken.Kind != SyntaxKind.OpenBracketToken)
+                return false;
+
+            using (this.GetDisposableResetPoint(resetOnDispose: true))
+            {
+                // Eat the `[`
+                EatToken();
+
+                // `[ id` could definitely begin an attribute.
+                if (this.IsTrueIdentifier())
+                    return true;
+
+                // `[ word: ...` could definitely start an attribute.
+                if (IsAttributeTarget())
+                    return true;
+
+                // If we see `[lit` (like `[0`) then this is def not an attribute, and should be parsed as a collection
+                // expr.  Note: this heuristic can be added to in the future.
+                if (SyntaxFacts.IsLiteralExpression(this.CurrentToken.Kind))
+                    return false;
+
+                return true;
```

Comment:
Is this basically an error recovery path? I thought if `[` is not followed by an identifier or attribute target it is bound to be invalid as an attribute.

Reply:
so this is error recovery, but attribute-parsing may still be prefered.  Consider something as simple as:

```c#
[
class C
```

We could parse this as a collection-expr as a top level statement, prior to a class.... or we can parse it as an attribute.

Right now we'll continue parsing that as an attribute, which is good here.  We can certain refine the above if we want.  But i really want to ensure large numeric arrays get better, without potentially regressing cases like this.